llvm/lib/Target/AIE/AIEBaseInstrInfo.h

```diff
@@ -84,6 +84,23 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
   virtual unsigned getCycleSeparatorOpcode() const {
     llvm_unreachable("Target didn't implement getCycleSeparatorOpcode");
   }
+  /// Return the opcode to be used for pushing a vector element at the LSB
+  /// position in a vector
+  virtual unsigned getGenericAddVectorEltOpcode() const { return -1; }
+  /// Return the opcode to be used for inserting a vector element at an
+  /// arbitrary position in a vector
+  virtual unsigned getGenericInsertVectorEltOpcode() const { return -1; }
+  /// Return the opcode to be used for extracting a vector element
+  /// \param signExt True if the extracted element shall be sign extended
+  virtual unsigned getGenericExtractVectorEltOpcode(bool SignExt) const {
+    return -1;
+  }
+  /// Return the opcode to be used for padding undefined values in the high bits
+  /// of a vector
+  virtual unsigned getGenericPadVectorOpcode() const { return -1; }
+  /// Return the opcode to be used for extracting a smaller vector by ignoring
+  /// the high bits
+  virtual unsigned getGenericUnpadVectorOpcode() const { return -1; }
```

Comment:
Nit: I feel we are really overloading InstrInfo with all those `getXXXXopcode()` function. At least for generic instrucitons, I'm wondering if we should not create e.g.
```
class AIEGenericInstructionsInterface {
  virtual unsigned getInsertVectorEltOpcode() const {...}
  virtual unsigned getExtractVectorEltOpcode() const {...}
  ...
```

I guess sometimes we will only really need to pass `AIEGenericInstructionsInterface`, and not the whole `AIEBaseInstrInfo`. This will also make unit testing easier, because we could just define a test interface for what is required to be tested.

What do you think?

Reply:
I don't really see an issue with adding more of these getter functions to InstrInfo, its a single place where we collect all architecture specific opcode information.

Now for AIE specific generic opcodes we could also consider them to be always available on every AIE architecture, and requiring them to be defined as the first opcodes after `GENERIC_OP_END`.
Then it doesn't matter which AIE namespace we use to access those opcodes (`AIE::` or `AIE2::` at this point).
We would have to be extremely careful not to introduce an opcode in between GENERIC_OP_END and the shared opcodes only for a single AIE version. Not sure that's any better

Reply:
The problem is that these "target specific generic machine instructions" do not exist in the `TargetOpcode` namespace, so we cannot translate them into the target namespace.

`TargetOpcode::` only contains the instructions up to `GENERIC_OP_END`, which also marks the beginning of the target specific (generic) machine instructions

Reply:
Yeah, but the idea could have a generic namespace that is an amalgamation of the different architectures. That has different drawbacks, like keeping it up to date and it becoming very large, but it is an alternative. Both are good to me and I happy with his as well